<!DOCTYPE html><html><head><meta name="viewport" content="width=device-width"/><meta charSet="utf-8"/><title>Mr.RS的个人博客-前端架构&amp;软实力-异常监控</title><meta name="next-head-count" content="3"/><link rel="preload" href="/next-blog/_next/static/css/1bf96eec8dc413e7.css" as="style"/><link rel="stylesheet" href="/next-blog/_next/static/css/1bf96eec8dc413e7.css" data-n-g=""/><noscript data-n-css=""></noscript><script defer="" nomodule="" src="/next-blog/_next/static/chunks/polyfills-5cd94c89d3acac5f.js"></script><script src="/next-blog/_next/static/chunks/webpack-cf185e4b59cdddcf.js" defer=""></script><script src="/next-blog/_next/static/chunks/framework-91d7f78b5b4003c8.js" defer=""></script><script src="/next-blog/_next/static/chunks/main-1c9e93b8392c1bc9.js" defer=""></script><script src="/next-blog/_next/static/chunks/pages/_app-ca6aae25bc99a05f.js" defer=""></script><script src="/next-blog/_next/static/chunks/899-1cd3a59abdb8ea18.js" defer=""></script><script src="/next-blog/_next/static/chunks/pages/post/%5Btitle%5D-406bad804ceffd1d.js" defer=""></script><script src="/next-blog/_next/static/p7mLHOZE5G1VHvh3Hmph3/_buildManifest.js" defer=""></script><script src="/next-blog/_next/static/p7mLHOZE5G1VHvh3Hmph3/_ssgManifest.js" defer=""></script><script src="/next-blog/_next/static/p7mLHOZE5G1VHvh3Hmph3/_middlewareManifest.js" defer=""></script></head><body><div id="__next" data-reactroot=""><header class="fixed top-0 py-4 left-0 right-0 h-auto px-10 shadow bg-white z-50"><nav><ul class="flex w-auto"><li class="mr-4 hover:text-yellow"><a href="/next-blog">首页</a></li><li class="mx-4 hover:text-yellow"><a href="/next-blog/categories">分类</a></li><li class="mx-4 hover:text-yellow"><a href="/next-blog/tags">标签</a></li><li class="ml-4 hover:text-yellow"><a href="/next-blog/timeline">归档</a></li></ul></nav></header><main class="mt-20 mb-4 mx-auto justify-center w-2/3"><div class="mx-auto lg:max-w-5xl md:max-w-3xl"><article class="flex flex-col justify-around pb-16 px-4 mx-auto space-y-10 text-base"><div class="flex flex-col space-y-4"><h1 class="inline pt-10 text-4xl text-skin-primary">前端架构&amp;软实力-异常监控</h1><div class="text-sm space-x-2 text-skin-muted"><span><svg fill="none" stroke="currentColor" viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg" class="w-5 h-5 inline-block mr-1"><path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M12 8v4l3 3m6-3a9 9 0 11-18 0 9 9 0 0118 0z"></path></svg>2020-06-08 22:33:01</span><span><svg fill="none" stroke="currentColor" viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg" class="w-5 h-5 inline-block mr-1"><path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M15 12a3 3 0 11-6 0 3 3 0 016 0z"></path><path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M2.458 12C3.732 7.943 7.523 5 12 5c4.478 0 8.268 2.943 9.542 7-1.274 4.057-5.064 7-9.542 7-4.477 0-8.268-2.943-9.542-7z"></path></svg>44</span></div></div><div class="mt-8 flex"><div class="w-full flex-auto custom-markdown-body"><div class="markdown-body" id="write"><h1 id="javascript中的异常处理方式">JavaScript中的异常处理方式</h1>
<p>对于JavaScript而言，我们面对的仅仅只是异常，异常的出现并不会直接导致JavaScript引擎崩溃，最多只会使当前执行的任务终止</p>
<ol>
<li>当前代码块将作为一个任务压入任务队列中，JavaScript引擎线程会不断地从任务队列中提取任务执行</li>
<li>当任务执行过程中出现异常，且异常没有捕获处理，则会一直沿着调用栈一层层向外抛出，最终终止当前任务地执行</li>
<li>JavaScript引擎线程继续从任务队列中取出下一个任务继续执行</li>
</ol>
<p>异常主要分为两大中：<strong>运行时错误</strong>和语<strong>法错误</strong>。语法错误在编写代码期间很容易被发现，运行时错误主要由取值/赋值异常、函数执行出错等</p>
<h1 id="如何捕获异常">如何捕获异常</h1>
<h2 id="捕获普通javascript异常">捕获普通JavaScript异常</h2>
<p>对于普通的JavaScript异常，常见的做法是使用<code>try-catch</code>进行包裹，当<code>try</code>中的代码块执行出错后，<code>catch</code>将能捕捉到异常，代码也可以继续执行下去</p>
<p>但<code>try-catch</code>处理异常的能力有限，<strong>只能捕捉到运行时非异步错误</strong>，对于语法错误和异步错误就显得无能为力，捕捉不到</p>
<h2 id="捕获异步javascript异常">捕获异步JavaScript异常</h2>
<p><code>window.onerror</code>捕获异常能力比<code>try-catch</code>稍微强点，无论是异步还是非异步运行时错误，<code>onerror</code>都能捕获到</p>
<pre><code class="hljs language-js"><span class="hljs-variable language_">window</span>.<span class="hljs-property">onerror</span> = <span class="hljs-keyword">function</span> (<span class="hljs-params">msg, url, row, col, error</span>) {
  <span class="hljs-variable language_">console</span>.<span class="hljs-title function_">log</span>(<span class="hljs-string">&#x27;我知道错误了&#x27;</span>);
  <span class="hljs-variable language_">console</span>.<span class="hljs-title function_">log</span>({
    msg,  url,  row, col, error
  })
  <span class="hljs-keyword">return</span> <span class="hljs-literal">true</span>;
};
</code></pre>
<p>使用注意：</p>
<ul>
<li>在实际的使用过程中，<code>windows.onerror</code> 主要是来捕获预料之外的错误，而 <code>try-catch</code> 则是用来在可预见情况下监控特定的错误，两者结合使用更加高效。</li>
<li><code>window.onerror</code> 函数只有在返回 true 的时候，异常才不会向上抛出，否则即使是知道异常的发生控制台还是会显示</li>
<li><code>window.onerror</code>无法捕获到网络异常的错误</li>
<li>无法捕获<code>Promise</code>异常</li>
</ul>
<h2 id="捕获promise-异常">捕获<strong>Promise</strong> 异常</h2>
<p>通过 <code>Promise</code> 可以帮助我们解决异步回调地狱的问题，但是一旦 ~实例抛出异常而你没有用 <code>catch</code> 去捕获的话，<code>onerror</code> 或 <code>try-catch</code> 也无能为力，无法捕捉到错误</p>
<p>对此可以添加一个 <strong><code>Promise</code> 全局异常捕获事件</strong> <code>unhandledrejection</code></p>
<pre><code class="hljs language-js"><span class="hljs-variable language_">window</span>.<span class="hljs-title function_">addEventListener</span>(<span class="hljs-string">&quot;unhandledrejection&quot;</span>, <span class="hljs-keyword">function</span>(<span class="hljs-params">e</span>){
  e.<span class="hljs-title function_">preventDefault</span>()
  <span class="hljs-variable language_">console</span>.<span class="hljs-title function_">log</span>(<span class="hljs-string">&#x27;我知道 promise 的错误了&#x27;</span>);
  <span class="hljs-variable language_">console</span>.<span class="hljs-title function_">log</span>(e.<span class="hljs-property">reason</span>);
  <span class="hljs-keyword">return</span> <span class="hljs-literal">true</span>;
});
</code></pre>
<p>|                    | 同步方法 | 异步方法 | 资源加载 | Promise | async/await |
| </p>
</div></div></div></article></div></main><footer class="flex flex-col items-center justify-center leading-8 text-sm w-full py-2 border-t text-skin-muted"><p>Designed &amp; developed by<a class="ml-1 text-yellow" target="_blank" href="https://github.com/mrrs878" rel="noreferrer">Mr.RS</a></p><p>©<!-- --> <!-- -->2022</p></footer></div><script id="__NEXT_DATA__" type="application/json">{"props":{"pageProps":{"post":{"title":"前端架构\u0026软实力-异常监控","tags":["异常监控"],"categories":"前端架构\u0026软实力","description":"# JavaScript中的异常处理方式\n\n对于JavaScript而言，我们面对的仅仅只是异常，异常的出现并不会直接导致JavaScript引擎崩溃，最多只会使当前执行的任务终止\n\n1. 当前代码块将作为一个任务压入任务队列中，JavaScript引擎线程会不断地从任务队列中提取任务执行\n2. 当任务执行过程中出现异常，且异常没有捕获处理，则会一直沿着调用栈一层层向外抛出，最终终止当前任务地","createDate":"2020-06-08 22:33:01","updateDate":"2021-02-20 17:15:07","body":"\u003ch1 id=\"javascript中的异常处理方式\"\u003eJavaScript中的异常处理方式\u003c/h1\u003e\n\u003cp\u003e对于JavaScript而言，我们面对的仅仅只是异常，异常的出现并不会直接导致JavaScript引擎崩溃，最多只会使当前执行的任务终止\u003c/p\u003e\n\u003col\u003e\n\u003cli\u003e当前代码块将作为一个任务压入任务队列中，JavaScript引擎线程会不断地从任务队列中提取任务执行\u003c/li\u003e\n\u003cli\u003e当任务执行过程中出现异常，且异常没有捕获处理，则会一直沿着调用栈一层层向外抛出，最终终止当前任务地执行\u003c/li\u003e\n\u003cli\u003eJavaScript引擎线程继续从任务队列中取出下一个任务继续执行\u003c/li\u003e\n\u003c/ol\u003e\n\u003cp\u003e异常主要分为两大中：\u003cstrong\u003e运行时错误\u003c/strong\u003e和语\u003cstrong\u003e法错误\u003c/strong\u003e。语法错误在编写代码期间很容易被发现，运行时错误主要由取值/赋值异常、函数执行出错等\u003c/p\u003e\n\u003ch1 id=\"如何捕获异常\"\u003e如何捕获异常\u003c/h1\u003e\n\u003ch2 id=\"捕获普通javascript异常\"\u003e捕获普通JavaScript异常\u003c/h2\u003e\n\u003cp\u003e对于普通的JavaScript异常，常见的做法是使用\u003ccode\u003etry-catch\u003c/code\u003e进行包裹，当\u003ccode\u003etry\u003c/code\u003e中的代码块执行出错后，\u003ccode\u003ecatch\u003c/code\u003e将能捕捉到异常，代码也可以继续执行下去\u003c/p\u003e\n\u003cp\u003e但\u003ccode\u003etry-catch\u003c/code\u003e处理异常的能力有限，\u003cstrong\u003e只能捕捉到运行时非异步错误\u003c/strong\u003e，对于语法错误和异步错误就显得无能为力，捕捉不到\u003c/p\u003e\n\u003ch2 id=\"捕获异步javascript异常\"\u003e捕获异步JavaScript异常\u003c/h2\u003e\n\u003cp\u003e\u003ccode\u003ewindow.onerror\u003c/code\u003e捕获异常能力比\u003ccode\u003etry-catch\u003c/code\u003e稍微强点，无论是异步还是非异步运行时错误，\u003ccode\u003eonerror\u003c/code\u003e都能捕获到\u003c/p\u003e\n\u003cpre\u003e\u003ccode class=\"hljs language-js\"\u003e\u003cspan class=\"hljs-variable language_\"\u003ewindow\u003c/span\u003e.\u003cspan class=\"hljs-property\"\u003eonerror\u003c/span\u003e = \u003cspan class=\"hljs-keyword\"\u003efunction\u003c/span\u003e (\u003cspan class=\"hljs-params\"\u003emsg, url, row, col, error\u003c/span\u003e) {\n  \u003cspan class=\"hljs-variable language_\"\u003econsole\u003c/span\u003e.\u003cspan class=\"hljs-title function_\"\u003elog\u003c/span\u003e(\u003cspan class=\"hljs-string\"\u003e\u0026#x27;我知道错误了\u0026#x27;\u003c/span\u003e);\n  \u003cspan class=\"hljs-variable language_\"\u003econsole\u003c/span\u003e.\u003cspan class=\"hljs-title function_\"\u003elog\u003c/span\u003e({\n    msg,  url,  row, col, error\n  })\n  \u003cspan class=\"hljs-keyword\"\u003ereturn\u003c/span\u003e \u003cspan class=\"hljs-literal\"\u003etrue\u003c/span\u003e;\n};\n\u003c/code\u003e\u003c/pre\u003e\n\u003cp\u003e使用注意：\u003c/p\u003e\n\u003cul\u003e\n\u003cli\u003e在实际的使用过程中，\u003ccode\u003ewindows.onerror\u003c/code\u003e 主要是来捕获预料之外的错误，而 \u003ccode\u003etry-catch\u003c/code\u003e 则是用来在可预见情况下监控特定的错误，两者结合使用更加高效。\u003c/li\u003e\n\u003cli\u003e\u003ccode\u003ewindow.onerror\u003c/code\u003e 函数只有在返回 true 的时候，异常才不会向上抛出，否则即使是知道异常的发生控制台还是会显示\u003c/li\u003e\n\u003cli\u003e\u003ccode\u003ewindow.onerror\u003c/code\u003e无法捕获到网络异常的错误\u003c/li\u003e\n\u003cli\u003e无法捕获\u003ccode\u003ePromise\u003c/code\u003e异常\u003c/li\u003e\n\u003c/ul\u003e\n\u003ch2 id=\"捕获promise-异常\"\u003e捕获\u003cstrong\u003ePromise\u003c/strong\u003e 异常\u003c/h2\u003e\n\u003cp\u003e通过 \u003ccode\u003ePromise\u003c/code\u003e 可以帮助我们解决异步回调地狱的问题，但是一旦 ~实例抛出异常而你没有用 \u003ccode\u003ecatch\u003c/code\u003e 去捕获的话，\u003ccode\u003eonerror\u003c/code\u003e 或 \u003ccode\u003etry-catch\u003c/code\u003e 也无能为力，无法捕捉到错误\u003c/p\u003e\n\u003cp\u003e对此可以添加一个 \u003cstrong\u003e\u003ccode\u003ePromise\u003c/code\u003e 全局异常捕获事件\u003c/strong\u003e \u003ccode\u003eunhandledrejection\u003c/code\u003e\u003c/p\u003e\n\u003cpre\u003e\u003ccode class=\"hljs language-js\"\u003e\u003cspan class=\"hljs-variable language_\"\u003ewindow\u003c/span\u003e.\u003cspan class=\"hljs-title function_\"\u003eaddEventListener\u003c/span\u003e(\u003cspan class=\"hljs-string\"\u003e\u0026quot;unhandledrejection\u0026quot;\u003c/span\u003e, \u003cspan class=\"hljs-keyword\"\u003efunction\u003c/span\u003e(\u003cspan class=\"hljs-params\"\u003ee\u003c/span\u003e){\n  e.\u003cspan class=\"hljs-title function_\"\u003epreventDefault\u003c/span\u003e()\n  \u003cspan class=\"hljs-variable language_\"\u003econsole\u003c/span\u003e.\u003cspan class=\"hljs-title function_\"\u003elog\u003c/span\u003e(\u003cspan class=\"hljs-string\"\u003e\u0026#x27;我知道 promise 的错误了\u0026#x27;\u003c/span\u003e);\n  \u003cspan class=\"hljs-variable language_\"\u003econsole\u003c/span\u003e.\u003cspan class=\"hljs-title function_\"\u003elog\u003c/span\u003e(e.\u003cspan class=\"hljs-property\"\u003ereason\u003c/span\u003e);\n  \u003cspan class=\"hljs-keyword\"\u003ereturn\u003c/span\u003e \u003cspan class=\"hljs-literal\"\u003etrue\u003c/span\u003e;\n});\n\u003c/code\u003e\u003c/pre\u003e\n\u003cp\u003e|                    | 同步方法 | 异步方法 | 资源加载 | Promise | async/await |\n| \u003c/p\u003e\n"}},"__N_SSG":true},"page":"/post/[title]","query":{"title":"前端架构\u0026软实力-异常监控"},"buildId":"p7mLHOZE5G1VHvh3Hmph3","assetPrefix":"/next-blog","isFallback":false,"gsp":true,"scriptLoader":[]}</script></body></html>